### DISCUSSION TITLE
Consistent `rows`/`numberOfLines` prop on both Android and iOS

### DISCUSSION THREAD

**Discussion by chrispader** [react-native-community/discussions-and-proposals#731] [Ideas]
# Feature Request

The `rows`/`numberOfLines` prop on the `TextInput` component is currently only available on Android. @Szymon20000 has been working on [a PR](https://github.com/facebook/react-native/pull/38021) that adds this prop on iOS and provides a consistent approach on how this prop affects the `TextInput`.

At the moment, on iOS there is no way to explicitly set a `TextInput` to a specific number of lines without manually setting the actual `height` in style. Creating a `TextInput` which should have a specific number of lines is therefore only possible, by somehow calculating the line height by font size, spacing, etc....

## Problems

The `rows`/`numberOfLines` props currently behave different on Android/Web and between different architectures as described [in this comment](https://github.com/facebook/react-native/pull/35703#issuecomment-1434708367).

Additionally, there are two props for the same use-case: `rows` and `numberOfLines`. We should reduce these to only one prop `rows`. There already is an ongoing discussion about replacing `numberOfLines` with `rows` completely: https://github.com/facebook/react-native/pull/35703#discussion_r1055825951

## Proposal

On all platforms, we want to have 2 props:

- `rows`: Sets the **fixed number of rows** on a `TextInput`.
  - The `TextInput` will always have these number of lines even if the content is less
  -  If the number of `rows` is higher than `maxRows`, it will be capped to the defined maximum
- `maxRows`: Sets the **maximum** number of rows/lines **up to which a `TextInput` can grow**.
  - If `rows` is not set, the `TextInput` will start with only 1 row and can grow up to the maximum number of rows
  - If `rows` is set, the `TextInput` will always have the defined rows.

> [!NOTE]  
> I'm ignoring `numberOfLines`/`maxNumberOfLines` props, as we want to remove them anyway

In [this comment](https://github.com/facebook/react-native/pull/34488#issuecomment-1226074712) @necolas also stated that we'll want to unify this prop on all platforms at some point.

### Internal testing

@cipolleschi mentioned [here](https://github.com/facebook/react-native/pull/38021#issuecomment-1691859464) that this change will affect internal (screenshot) testing. I think in order to achieve this interface, we'll eventually want to update the internal tests. Up to this point though, we can also instead name these props `displayedRows` and `maxDisplayedRows`, to prevent internal tests from failing. 

## What's next?

@Szymon20000 has already [implemented this on iOS](https://github.com/facebook/react-native/pull/38021) and also [improved the Android code](https://github.com/facebook/react-native/pull/35703). We'll want to continue reviews on the first (iOS) PR and possibly split it up even more (between Fabric and Paper?). 

---

friendly bump @Szymon2000 @roryabraham @cipolleschi @necolas

**Comment by cipolleschi** [comment]
As stated, the two existing props have different behaviors across platforms.
We cannot reuse them: 
- this will require too much work for any Meta engineer to update the internal system
- we cannot outsource that to some external contributor because they won't have access to our code base
- we don't have capacity to follow that effort.

On top of those practical things, changing the existent behavior of existing props is a huge breaking change. It will not only break MEta code, but every other app that uses those properties expecting the current behavior.

Personally, I'm okay with a flow like this:
- create an (or two) additional props with the desired behavior.
- if we think that the behavior of old props is wrong, we can deprecate them.

Disclaimer: I will not have capacity to follow this effort. Somebody else @ Meta should champion this.

**Reply by chrispader** [reply]
I understand. Do you or anyone else at meta agree with the props `displayedRows` and `maxDisplayedRows` instead? These props would not conflict with the already existing props, but rather just implement our desired behaviour.

Who might be able to chime in here? maybe @NickGerleman :)

**Reply by jamonholmgren** [reply]
Rather than having different property names that are not aligned with web, I think I’d rather that there were a new TextInput component that wraps the current TextInput but standardizes behavior going forward. Perhaps it could just be called `TextArea` to further align with web. It would largely be compatible with TextInput but otherwise work more consistently.

Prior art: [Tamagui Textarea](https://tamagui.dev/docs/components/inputs/1.0.0)

**Reply by cipolleschi** [reply]
I like @jamonholmgren idea a lot.
This would also make the new component "independent" from the React Native core. It can be a separate package that you can use if you need this specific behavior.
You might start by copying the current component, add the behavior that you have in your PR, and publish it as a separate library. 
Then replace all the JS calls from `<TextInput>` to `<YourComponent>` and that's basically done.

**Reply by chrispader** [reply]
Hmm, i'm personally okay with having this functionality in a separate component to avoid conflicts with other codebases, but it would be cool to ship and maintain this alongside the `react-native` core library.

**Reply by cipolleschi** [reply]
Why, If I can ask?
Historically, there have been an effort called lean-core, for which we are moving out of the core those extra components. Bringing a new one in, for a feature we don't use internally, seem in direct opposition to that effort and to the direction of the framework. 🤔

**Reply by fabOnReact** [reply]
@cipolleschi regarding Meta effort for lean-core for Text/TextInput, would they be open to add more types for third-party libraries that want to work with raw text as children prop? thanks a lot (related Issue https://github.com/facebook/react-native/issues/25136)

https://github.com/fabriziobertoglio1987/react/blob/96740fbff5de90b3899db8d9c7df668c675c632b/packages/react-native-renderer/src/ReactNativeHostConfig.js#L200-L210

https://github.com/fabriziobertoglio1987/react-native/blob/385473522cbc525aad08500f5a752dea734c14c3/packages/react-native/Libraries/Renderer/implementations/ReactNativeRenderer-dev.js#L5373

**Reply by cipolleschi** [reply]
Uhm, good question.
I don't have much context on this, but perhaps @sammy-SC or @javache can shed more light on why this is happening and also if that is still true for Fabric.

In the [Fabric example from the documentation](https://reactnative.dev/docs/0.71/the-new-architecture/pillars-fabric-components#ios) it is possible to render some text without using a Text component or any one specified in the HostConfig.

**Reply by sebazelonka** [reply]
My 2 cents here. Having TextArea as part of lean-core is the way to go, as Textarea is an everyday use case and part of the web platform. It is the expected behavior when you need something like that on your app.
